Modernize util/strencodings and util/string: `string_view` and `optional`

sipa opened the pull request:
Make use of `std::string_view` and `std::optional` in the util/{strencodings, string} files.

This avoids many temporary string/vector objects being created, while making the interface easier to read. Changes include:
* Make all input arguments in functions in util/strencodings and util/string take `std::string_view` instead of `std::string`.
* Add `RemovePrefixView` and `TrimStringView` which also *return* `std::string_view` objects (the corresponding `RemovePrefix` and `TrimString` keep returning an `std::string`, as that's needed in many call sites still).
* Stop returning `std::string` from `DecodeBase32` and `DecodeBase64`, but return vectors. Base32/64 are fundamentally algorithms for encoding bytes as strings; returning `std::string` from those (especially doing it conditionally based on the input arguments/types) is just bizarre.
* Stop taking a `bool* pf_invalid` output argument pointer in `DecodeBase32` and `DecodeBase64`; return an `std::optional` instead.
* Make `DecodeBase32` and `DecodeBase64` more efficient by doing the conversion from characters to integer symbols on-the-fly rather than through a temporary vector.

sipa committed:
Make ParseHex use string_view

sipa committed:
Make IsHex use string_view

sipa committed:
Make IsHexNumber use string_view

sipa committed:
Make SanitizeString use string_view

sipa committed:
Generalize ConvertBits to permit transforming the input

sipa committed:
Make DecodeBase{32,64} always return vector, not string

Base32/base64 are mechanisms for encoding binary data. That they'd
decode to a string is just bizarre. The fact that they'd do that
based on the type of input arguments even more so.

sipa committed:
Make DecodeBase{32,64} return optional instead of taking bool*

sipa committed:
Make DecodeBase{32,64} take string_view arguments

sipa committed:
Use std::string_view throughout util strencodings/string

laanwj commented:
Concept ACK

laanwj commented on src/util/strencodings.cpp:
```diff
@@ -58,56 +58,43 @@ signed char HexDigit(char c)
 }
 
-std::vector<unsigned char> ParseHex(const char* psz)
+std::vector<unsigned char> ParseHex(std::string_view str)
```
ParseHex itself could use error feedback using an `optional` at some point too. But this will need revisiting all call sites, too much for this PR.

sipa replied:
Right, good for a follow-up.

maflcko reviewed (approved):
> Returning a string_view can be somewhat dangerous (in combination with auto? also when the original string goes out of scope, especially if it's a temporary), so giving these functions new names is a good idea.

A new name is good, but doesn't solve the problem. I think this can be fixed with an attribute: https://github.com/bitcoin/bitcoin/issues/20493


I think this can be merged, but I left a few nits and questions.


review ACK 6e1f14780452d956503f1dbc3c91b17b44287f77 💢

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 6e1f14780452d956503f1dbc3c91b17b44287f77 💢
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUghZAv+JPOBXiEZ9L1jCF9HPLDO1vGLEuvq6hw6EkaMSKd0Zn3UaiYvP98jPBMZ
HzTidM3yPWwDfLw3QVTlPCdt0J8wBAwOg5ZgjGRZ7fu1hn0SZ93q960giU68DbeJ
ZZCF4fAYPOkTqf1a06DnV7anglhEIG/JXMg+XESTvUGmtdKnP48sdwek4NZOw1h7
PcA+Ck38qlb68W38HQMwsQYr/aO2Ftc2Utw2GMv3Kkr7nD6kIwtuH9WOTWc8Kw2H
vr5NeRfQS1smm9rP+VrixKmCaC1h37r6NMuXDqh/MRNj77ONwmLDLWES/4bP9ioc
w814fQled+vIue+Mv5yBUNeGZmlOryQsC7wJwNkiRNKxG5H+bSn2KUKEtZAGjHpv
rfNI9X9nbDCvDQdHemrYCessHaCfGFCMeUDC4PWDZ0uNaZSC6RrU0y8pESeMOzxe
BJJxSpq0sHCOAVt8LKGyumBlma/k2+n9d6UNo3++K/am9Uxei4kFgLJ0QebpXKqm
NqMH+EA3
=/7qh
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented on src/test/base32_tests.cpp:
```diff
@@ -22,8 +22,8 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
-        std::string strDec = DecodeBase32(vstrOut[i]);
-        BOOST_CHECK_EQUAL(strDec, vstrIn[i]);
+        auto dec = DecodeBase32(vstrOut[i]);
+        BOOST_CHECK_MESSAGE(MakeByteSpan(dec) == MakeByteSpan(vstrIn[i]), vstrIn[i]);
```
edit: Looks like the binary data in this test are printable-ascii-only strings, so feel free to ignore. Also it looks like boost doesn't attempt to sanitize the binary data anyway.

Though, I still slightly prefer to print both the decoded string and the input string on failure, as opposed to just the input string.

sipa replied:
I've changed them to just reporting the encoded string in case of failure; that should still be sufficient to notice the offending test case, and keeps working if we'd add cases that have non-printable input. Ok?

MarcoFalke commented on src/util/strencodings.cpp:
```diff
@@ -144,40 +144,23 @@ std::optional<std::vector<unsigned char>> DecodeBase64(const char* p)
+    if (str.size() % 4 != 0) return {};
+    /* One or two = characters at the end are permitted. */
+    if (str.size() >= 1 && str.back() == '=') str.remove_suffix(1);
+    if (str.size() >= 1 && str.back() == '=') str.remove_suffix(1);
```
If true, it could make sense to mention this behaviour change in the commit description (or pull request description).

sipa replied:
Interesting! The answer is that it's complicated.

So, the test code you suggest here actually passes, but only because it isn't checking the invalid output flag. Current master, I believe, decodes input which isn't a multiple of 4 characters correctly (as if it had padding) but does set the `*pf_invalid` flag.

So this is indeed a behavior change, but only for call sites which don't check the invalid flag. I think there is only one such site: `RPCAuthorized()` in httprpc.cpp, and I think that's a bug (the HTTP spec RFC 7617 for Basic authentication specifies the use of padded Base64 on very casual reading).

If so, the commit that switches to `std::optional`-wrapped results actually fixes this.

Thoughts?

MarcoFalke replied:
Oh right. The commit that changes behaviour is likely the std::optional one (889a5f107c5187c9232f340adfaedfe1b90361d5), not the string_view one I mentioned earlier.

I guess there is nothing needed to be done here. I just wanted to mention the behaviour change here, so that we aren't surprised when someone creates a bug report "My broken javascript rpc library doesn't work".

sipa replied:
Moved the behavior-changing aspect to a separate commit.

laanwj commented:
> A new name is good, but doesn't solve the problem. I think this can be fixed with an attribute: https://github.com/bitcoin/bitcoin/issues/20493

Agree, I just meant that putting a string_view-returning function as a drop-in replacement would have been a very bad idea, a new function with a new name means that people using it can reason about the risks anew. Sure, if an attribute can help avoid such problems more widely that's of course even better.

sipa commented on src/util/strencodings.cpp:
```diff
@@ -81,32 +81,24 @@ bool IsHexNumber(const std::string& str)
+        }
+        auto c1 = HexDigit(*(it++));
+        auto c2 = HexDigit(*(it++));
+        if (c1 == (signed char)-1 || c2 == (signed char)-1) break;
```
I left the casts there because the original calls had them too. I believe they are indeed redundant here, I'll remove them.

sipa replied:
Replaced it with `< 0` checks.

sipa commented on src/test/base64_tests.cpp:
```diff
@@ -19,8 +19,8 @@ BOOST_AUTO_TEST_CASE(base64_testvectors)
-        std::string strDec = DecodeBase64(strEnc);
-        BOOST_CHECK_EQUAL(strDec, vstrIn[i]);
+        auto dec = DecodeBase64(strEnc);
+        BOOST_CHECK_MESSAGE(MakeByteSpan(dec) == MakeByteSpan(vstrIn[i]), vstrIn[i]);
```
See https://github.com/bitcoin/bitcoin/pull/24764#discussion_r842806254

sipa commented on src/test/base32_tests.cpp:
```diff
@@ -23,19 +23,14 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
         BOOST_CHECK_EQUAL(strEnc, vstrOutNoPadding[i]);
         auto dec = DecodeBase32(vstrOut[i]);
-        BOOST_CHECK_MESSAGE(MakeByteSpan(dec) == MakeByteSpan(vstrIn[i]), vstrIn[i]);
+        BOOST_CHECK_MESSAGE(dec && MakeByteSpan(*dec) == MakeByteSpan(vstrIn[i]), vstrIn[i]);
```
Used `BOOST_REQUIRE(dec);`.

sipa commented on src/util/strencodings.cpp:
```diff
@@ -244,20 +239,15 @@ std::vector<unsigned char> DecodeBase32(const char* p, bool* pf_invalid)
-        return {};
-    }
-    return DecodeBase32(str.c_str(), pf_invalid);
+    if (!ValidAsCString(str)) return {};
```
Done.

MarcoFalke commented:
review ACK e4092c5230bdd1f6439c70fa22b68381e5cc6899 🐤

Only changes:
* Remove confusing (signed char) cast
* Use shorter str.substr(0, 2) == "0x" over manual matching/implementation
* A minor change in a unit test
* Whitespace changes

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e4092c5230bdd1f6439c70fa22b68381e5cc6899 🐤

Only changes:
* Remove confusing (signed char) cast
* Use shorter str.substr(0, 2) == "0x" over manual matching/implementation
* A minor change in a unit test
* Whitespace changes
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjhQQwAsEPZBkw/LyKgYc/xITU1DL5Oausc0jYNBH7WrBEh055OyfUZ5NgxM87K
bcJTNQMCmLVfWnOcg0FUELaXCGbqua8qPgiKwQT5nxlHN0DXpakSSWrkGh2HKIGe
zwPKwNrMEH7TrCs3Tkn3ZTkiHI3gl+AFTL5zdw1b/sMwPapeZLDqoAQZXTqj+LRB
0d44jHX6oO5faHRI6p+fErVtg40Q+3s49CkYXbTl/u1HEWaKcWBxk0wZ9JHUUWuY
Z2K1OWDWH/KuR/gYlUUuVfJk5jgG6NqNgzErLUjCtYjQ6LGpI90e7ud/ahv2Xndb
vRzi/r/XAQrzlYLKM4ZuoeiTTSKxLJHD6353KVgPVKETioM6o1eyn3qWpg5JwnT4
VglcsozZzoVbSdghoQ397NhFjI+Lfs+d60EcUB8lzPgzP00gAo2smlXXM0QMBRV/
XOweIVcmYulLYK0LnBXbQy4GSqyE06JsIKpugwFJEcKb7mkdJSdJ8Xg4xxwwYyjR
CdV7knXn
=yyjx
-----END PGP SIGNATURE-----
```


</details>

Empact commented:
Concept ACK - the `ConvertBits` functionality is unused in this PR - maybe it could be saved for a time when it is needed?

nits:
* Mild preference for more strict separation between refactoring and behavior modification - would make this easier to ACK, at least in part
* There is a check in `IsHex` (`% 2 == 0`) that could be applied to `IsHexNumber`
* Could inline or return early in some cases where you're converting to return `optional` (previous implementation was constrained by the need to set the val) 889a5f107c5187c9232f340adfaedfe1b90361d5

sipa commented:
> Mild preference for more strict separation between refactoring and behavior modification - would make this easier to ACK, at least in part

Yeah, I'll separate them (now that we figured out the issue with ignoring the pfInvalid flag in the http auth logic).

> There is a check in IsHex (% 2 == 0) that could be applied to IsHexNumber

That would be incorrect. Hex numbers don't need to be a multiple of 2 characters (0xF is a valid hex number, e.g.).

sipa committed:
Reject incorrect base64 in HTTP auth

In addition, to make sure that no call site ignores the invalid
decoding status, make the pf_invalid argument mandatory.

maflcko reviewed (approved):
re-ACK 5ea072ed08b43cdfdf8554597dcf8c1a51ef4dae only effective change is adding a "return false" 🌷

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 5ea072ed08b43cdfdf8554597dcf8c1a51ef4dae only effective change is adding a "return false" 🌷
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgUyQwAhIAVoAgFscYNynT/fORwnmC2JJczY8Zne+7IKHyQadrQGB3uGARyJceI
WDRjN7aLHHJwFmFZ/Lz4HGa5sZYwAUC/Vz9y2eX0nu2PInfZgdtNbx8JT0cx5wkX
03LWXiek/qiy6BeEctpE1nP37iSAqB4E18yEhCMqdLe+m7KCq/rTZN2XdF/R3DHK
j9mqyyy7M5dYQUE2dViBdqTydSrhGIhw2HqHXytP8/urfFB04wLXeJHNDZ6lRfZ1
zq28KVL2FOuFXdYTIAtdeCAFAXFgwiZ2lYik1+aftCBBO9BdBnmKqRD0/ERAC2u8
TfVPFMQECHbWxcxGwZGBjhNhcXGI2fVUF2g4SqJVssqPnG53GcSRvN6H0tcKDAH7
g1x6fqDibKBIUXcrTp8l5tlZYknN2wU0wq9M7U6uJIt5gGWOTlpYpbQsMIlzg6Ls
omRKFXjTXHvFmO8eilHMaMVk6qY/12MjoTyHZ1JwN7TWHRxH6Q9XNTFeq1VJE/qd
uqaqLwJe
=bejp
-----END PGP SIGNATURE-----
```


</details>

martinus commented on src/util/string.h:
```diff
@@ -75,9 +85,12 @@ inline std::string MakeUnorderedList(const std::vector<std::string>& items)
  * Check if a string does not contain any embedded NUL (\0) characters
  */
-[[nodiscard]] inline bool ValidAsCString(const std::string& str) noexcept
+[[nodiscard]] inline bool ValidAsCString(std::string_view str) noexcept
```
I think the name `ValidAsCString` is now a bit dangerous. I'd assume that it returns `true` when the argument is 0 terminated at the end (which was the case for the `const std::string& str` argument), but with a `std::string_view` argument this might not be the case.

So now even when `ValidAsCString(str)` returns `true`, it is not allowed to use e.g. `strlen(str.data())`.

I would rename the method to something like `ContainsNoNUL` to make it more clear what's checked, or keep the `const std::string&` argument for now

laanwj replied:
According to the documentation, it only checks if a string has embedded `\0` characters. That std::string is zero-terminated internally is an implementation detail, the caller is still responsible for passing `c_str()` (and never `.data()`) when passing it to a function that takes a C string.

martinus replied:
well, maybe it's just me, but from the interface `ValidAsCString(std::string_view str)` I'd guess that it returns `true` when it actually finds a `\0` somewhere, but it's exactly the opposite :slightly_smiling_face:

MarcoFalke replied:
I have no objection to adding a scripted-diff to rename `ValidAsCString` to `ContainsNoNUL`. Either in this pull or as a follow-up.

MarcoFalke replied:
@sipa

MarcoFalke replied:
@martinus See https://github.com/bitcoin/bitcoin/pull/25001

MarcoFalke commented:
I am happy to pick this up if you are no longer working on it

sipa commented:
@MarcoFalke Go for it.
